异常就是程序出现了问题。

- Error：代表出现了系统级的问题（属于严重的问题），系统一旦出现问题，将会被封装成Error对象，与程序员无关。
- Exception：异常，代表程序可能出现的问题。我们通常使用Exception以及其他的子类来封装程序出现的问题
  - 运行时异常：RuntimeException及其子类，编译阶段不会出现异常提醒，但运行时会出现异常（如：数组引发的越界）
  - 编译时异常：编译阶段就会出现异常提醒的。（如：日期解析异常）



# 异常的作用

- 作用一：异常是用来查询bug的关键参考信息
- 作用二：异常可以作为方法内部的一种特殊返回值，以便通知调用者底层的执行情况



# 异常的处理方式

## JVM默认的处理方式

- 把异常的名称，异常原因及异常出现的位置等信息输出在控制台
- 程序停止执行，下面的代码也不会再执行了

## 自己处理（捕获异常）

- ```java
  try {
    可能出现的异常代码;
  } catch(异常类名 变量名) {
    异常的处理代码;
  }
  ```

- 当代码出现异常时，可以让程序继续往下运行

- try块中出现了异常，程序会创建一个异常对象

- 程序会用异常对象与catch()中的类进行对比，看catch中的变量能否接收这个对象

- 如果能被接收，就表示该异常就被捕获，执行catch中的代码

- 当catch中的所有代码执行完毕，继续执行try...catch体系下面其他的代码

### 如果try中没有遇到问题，如何执行？

会把try中的代码全部执行完毕，不会执行catch中的代码。**只有当出现了异常，才会执行catch中的代码。**

### 如果try中遇到了多个问题，怎么执行？

写多个catch进行捕获

```java
try {
  
} catch() {
  
} catch() {
  
}
```

**注意：Exception必须写在最下面的catch()中，否则会报错。**因为Exception是所有错误的父类，放在前面所有错误会直接全部匹配上Exception，下面的catch()块全部没用了。

JDK 7以后有了新的语法：

```java
try {
  
} catch(ArrayIndexOutOfBoundsException | ArithmeticException e) {
  
}
```

### 如果try中的问题没有被捕获怎么办

catch接不住的话，就会执行JVM默认的执行方式，直接在控制台打印输出异常。

### 如果try中某行出现了问题，那么这一行下面的代码还会执行吗？

不会，**会直接离开try块，跳转到对应的catch中**



## 抛出异常

### Throws

写在方法定义出，表示声明一个异常。告诉调用者，使用本方法可能会有哪些异常。

```java
public void 方法() throws 异常类名1, 异常类名2... {
  
}
```

编译时异常：必须要写。

运行时异常：可以不写。



### Throw

写在方法内，结束方法手动抛出异常对象，交给调用者处理。方法中下面的代码不再执行了。

# 异常中的常见方法

## Trowable的成员方法

| 方法名称                    | 说明                            |
| --------------------------- | ------------------------------- |
| public String getMessage()  | 返回此throwable的详细消息字符串 |
| public String toString()    | 返回此可抛出的简短描述          |
| public void printStackTrace | 把异常的错误信息输出在控制台    |



# 自定义异常

1. 定义异常类
2. 写继承关系
3. 空参构造
4. 带参构造

